### PR TITLE
Improved heartbeat implementation, redo actions, and unit tests

### PR DESCRIPTION
**Pull request by Osmose**
This is a rather large PR. It does several things:
- Changes how we define and register actions. Actions are now classes with an async `execute` function that implements what the action actually does. Instead of an `arguments.json` file next to the implementation, we now have a `package.json` file for each action that defines the entry point as well as the argument schema it will use.
  
  The change to a class was to help group properties needed across functions in actions so that they don't interfere with eachother, as well as to aid testing them. The change to a `package.json` file was to support adding a new `driverVersion` field for future use with pinning actions to a specific version of the Normandy driver.
- Adds Sphinx documentation for the "driver API". Since self-repair and the add-on (and possibly others) will be handing this driver object to actions, we need one place where the API is documented and versioned. This allows actions to declare a version compatibility with the documented API, and allows execution environments to avoid running actions that aren't compatible with the driver they expose.
- Adds unit tests for the actions. Soon we should run these in Travis/Circle, but there's already enough stuff in this PR.

There's a few things I'd like to call out for extra scrutiny:
- The tests have a lot of boilerplate. I tried using some model makers to help cut it down but it's still a bit more verbose than I'd like. Suggestions on improvements would be great.
- The versions on the docs vs the driver versions in the manifests don't match up currently because it's not clear what will change before we actually put these in prod. Also the version enforcement doesn't have to function until we have more than one version out, so it shouldn't be an issue right now.

@brittanystoroz @mythmon r?

### CODE REVIEW

**Comment by Osmose** [comment]
Oh whoops, I should probably document how to run the tests and build the docs. Will add another commit with that.

**Comment by Osmose** [comment]
Docs added. They're pretty minimal, but I hope to get them in better shape at the start of Q2 before we dive into feature dev.

File: actions/utils.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ */
+export function weightedChoose(choices) {
+    // Rolling sums of weights; each weight is the choice's weight plus
+    // the weight of all previous choices.
+    let weights = [choices[0].weight];
+    for (let choice of choices.slice(1)) {
+        weights.push(choice.weight + weights[weights.length - 1]);
+    }
+
+    let choice = Math.random() * weights[weights.length - 1];
```
Comment (brittanydionigi) [inline actions/utils.js]: I would maybe change the name of this variable from `choice` to something like...`weightLimit`? `chosenWeight`? `randomWeight`?

It's a little hard to follow this code with the current naming...`choices` represents an array of objects that each have a weight property, but `choice` is a random integer weight we're using to determine `chosen`. (Unless you're inside the for loop on line 24, in which case `choice` is an object with a weight property.)

**Comment by brittanydionigi** [comment]
This looks good so far and makes sense to me. I was able to get everything working locally although I did run into this when pip installing Sphinx. Is this just me? I wound up pip installing all those deps manually.

```
In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:
    docutils>=0.11 from https://pypi.python.org/packages/3.4/d/docutils/docutils-0.12-py3-none-any.whl#md5=cd86e38514d9032f0a6be8dd08d3818b (from Sphinx==1.3.6->-r docs/requirements.txt (line 1))
    alabaster<0.8,>=0.7 from https://pypi.python.org/packages/py2.py3/a/alabaster/alabaster-0.7.7-py2.py3-none-any.whl#md5=2a6a0e37648fe9eb60882e6b8afdc27f (from Sphinx==1.3.6->-r docs/requirements.txt (line 1))
    Pygments>=2.0 from https://pypi.python.org/packages/2.7/P/Pygments/Pygments-2.1.3-py2.py3-none-any.whl#md5=242f73df7fa160366495a8f11f816e02 (from Sphinx==1.3.6->-r docs/requirements.txt (line 1))
    Jinja2>=2.3 from https://pypi.python.org/packages/2.7/J/Jinja2/Jinja2-2.8-py2.py3-none-any.whl#md5=75acb6f1abfc46ed75f4fd392f321ac2 (from Sphinx==1.3.6->-r docs/requirements.txt (line 1))
```

**Comment by Osmose** [comment]
> Is this just me?

Nope I'm dumb and didn't test pip installing that stuff like a dummy.

File: actions/show-heartbeat/index.js [hunk 2]
```diff
@@ -1,20 +1,90 @@
...
-    }).then(() => {
-        console.log('Heartbeat happened!');
-    });
-});
+import 'babel-polyfill';
+
+import {Action, registerAction, weightedChoose} from '../utils';
+
+const VERSION = 52; // Increase when changed.
+const LAST_SHOWN_DELAY = 100 * 60 * 60 * 24 * 7; // 7 days
```
Comment (mythmon) [inline actions/show-heartbeat/index.js]: Should this 100 be 1000? Or do you actually want centiseconds here?

File: actions/show-heartbeat/index.js [hunk 3]
```diff
@@ -1,20 +1,90 @@
...
+    }
+
+    async annotatePostAnswerUrl(url) {
+        let appInfo = await this.normandy.getAppInfo();
+        let params = (
+            'source=heartbeat' +
+            '&surveyversion=' + VERSION +
+            '&updateChannel=' + appInfo.defaultUpdateChannel +
+            '&fxVersion=' + appInfo.version
+        );
```
Comment (mythmon) [inline actions/show-heartbeat/index.js]: This situation sucks. I wish there were better things.

I probably would have written it as something like 

``` js
let args = [
  ['source', 'heartbeat'],
  ['surveyversion', VERSION],
  ['updateChannel', applInfo.defaultUpdateChannel],
  ['fxVersion', appInfo.version],
];
let params = args.forEach(([a, b]) => `${a}=${b}`).join('&');
```

This way the syntax and punctuation isn't part of the string, and is harder to screw up. If JS had something like Python's `{}.items()`, this could be even better.

File: actions/show-heartbeat/index.js [hunk 4]
```diff
@@ -1,20 +1,90 @@
...
+     *                           object.
+     * @param  {object} defaults Default values for survey attributes if they aren't
+     *                           specified.
+     * @return {object}          The chosen survey, with the defaults applied.
+     */
+    chooseSurvey(surveys, defaults) {
+        let finalSurvey = Object.assign({}, weightedChoose(surveys));
+        for (let prop in finalSurvey) {
+            if (!finalSurvey[prop]) {
+                finalSurvey[prop] = defaults[prop];
```
Comment (mythmon) [inline actions/show-heartbeat/index.js:86]: I think you want to loop through the props of `defaults`, not the props of `finalSurvey`.

File: actions/utils.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    let chosen = choices[choices.length - 1];
+    for (let k = 0; k < weights.length - 1; k++) {
+        if (choice < weights[k]) {
+            chosen = choices[k];
+            break;
+        }
+    }
+
+    return chosen;
```
Comment (mythmon) [inline actions/utils.js]: I think you can write a simpler version of this:

``` js
let sumWeight = 0;
for (let choice of choices) {
    sumWeight += choice.weight;
}
let chosenWeight = Math.random() * maxWeight;

for (let choice of choices) {
    chosenWeight -= choice.weight;
    if (chosenWeight <= 0) {
        return choice;
    }
}
```

File: docs/make.bat [hunk 6]
```diff
@@ -0,0 +1,263 @@
+@ECHO OFF
```
Comment (mythmon) [inline docs/make.bat]: I wouldn't have committed this file.

File: docs/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+Sphinx==1.3.6 \
+    --hash=sha256:f26cabf24fe553b3dd7935a9e730cb570d64147e053fc133675638fa576df467 \
+    --hash=sha256:0b3be9221220af36090d139192347bb5f42b13401e4c53a4de8f5329a7a3ff8a
```
Comment (mythmon) [inline docs/requirements.txt]: Does RTD have pip8 now? Or should we leave off the hashes for now?

File: test/show-heartbeat.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        let action = new ShowHeartbeatAction(this.normandy, recipe);
+
+        this.normandy.mock.storage.data['lastShown'] = '100';
+        this.sinon.stub(Date, 'now').returns(10);
+
+        await action.execute();
+        expect(this.normandy.showHeartbeat).to.not.have.been.called;
+    });
+
+    it('should show heartbeat if it hasn\'t shown within the past 7 days', async function() {
```
Comment (mythmon) [inline test/show-heartbeat.js]: Nit: Strings with single quotes in them should use double quotes instead of escaping single quotes, where possible.

File: actions/show-heartbeat/index.js (hunk 2 above)
Comment (Osmose) [inline actions/show-heartbeat/index.js]: shhhhhhhhhhhh

File: actions/show-heartbeat/index.js (hunk 3 above)
Comment (Osmose) [inline actions/show-heartbeat/index.js]: I _really_ wish jsdom supported `URLUtils.searchParams`. I hate that we can't use a feature available in Firefox due to it not being mocked in our tests, but I don't think we have time to solve that right now.

File: docs/requirements.txt (hunk 7 above)
Comment (Osmose) [inline docs/requirements.txt]: Why isn't there a rage option in this reaction list

**Comment by Osmose** [comment]
Updated with fixes from feedback.

**Comment by Osmose** [comment]
Added a new commit that removes `babel-polyfill` from being included with each action. Instead I'm going to have the self-repair page include `babel-polyfill` automatically. This avoids an issue I ran into where having more than one recipe that used `babel-polyfill` failed since you can only include it once on a page.

In Q2 we'll figure out a better way to share code among actions (and we discussed a few options early on in the project) but for now this will let us ship.

**Comment by Osmose** [comment]
Aaand two more commits (I'm working on the patch to the server to support these new actions and running into changes that need to be made.)

**Comment by brittanydionigi** [comment]
all recent fixes lgtm :tada:

**Comment by mythmon** [comment]
r+
